### PR TITLE
rho analyzer and empty area correction added to customize

### PR DESCRIPTION
**Pull request by dgulhan**
This PR includes rho and a correction for rho to take into account empty areas, both of these producers are based on fastjet algorithms. The reconstruction sequence is modified to run these producers and keep them in the output. There is a fix in the BuildFile for the issue discussed here:
https://github.com/cms-sw/cmssw/issues/14862#issuecomment-226190302
These producers were initially developed and tested for PbPb collisions and pp collisions with HI reconstruction in MB and dijet embedded samples in CMSSW_7_5_8_patch3, here are some links to validation plots:
https://twiki.cern.ch/twiki/pub/CMS/HiJetReco2015/160317_RCPeripheralCorr.pdf
https://twiki.cern.ch/twiki/pub/CMS/HiJetReco2015/160317_RCPeripheralCorr.pdf
In pp collisions a small dijet sample was produced with 112 option of the matrix with additional customizations for pPb to test the producers in CMSSW_8_1_X with pp reconstruction.
We would like to merge it to the IB in order to use it for 2016 pPb runs planned to be reconstructed with CMSSW_8_0, and future PbPb runs.

@mverwe @mandrenguyen @yetkinyilmaz

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

@dgulhan 
based on a quick scan of the changes, there will be quite a few comments in the code review

**Comment by dgulhan** [comment]
There was a missing python file in the PR, the tests succeeded despite this probably because --customise  RecoHI/Configuration/customise_PPwithHI.customisePPwithHI option was not used in the tests. Without this option the changes are not used.

**Comment by cvuosalo** [comment]
@cmsbuild please test

File: RecoHI/HiJetAlgos/plugins/HiFJGridEmptyAreaCalculator.h [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+
+#include "TMath.h"
+
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidate.h"
+#include "DataFormats/PatCandidates/interface/Jet.h"
+
+class HiFJGridEmptyAreaCalculator : public edm::EDProducer {
```
Comment (cvuosalo) [inline RecoHI/HiJetAlgos/plugins/HiFJGridEmptyAreaCalculator.h]: Please change to inherit from a thread-safe EDProducer. More info here:

https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkModuleTypes#Stream_Module

File: RecoHI/HiJetAlgos/plugins/HiFJRhoProducer.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#include "FWCore/Framework/interface/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "DataFormats/JetReco/interface/Jet.h"
+
+//
+// class declaration
+//
+
+class HiFJRhoProducer : public edm::EDProducer {
```
Comment (cvuosalo) [inline RecoHI/HiJetAlgos/plugins/HiFJRhoProducer.h]: Change EDProducer for thread-safety.

File: RecoHI/HiJetAlgos/python/hiFJGridEmptyAreaCalculator_cff.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+                                 bandWidth = cms.untracked.double(0.2),
+                                 mapEtaEdges = cms.InputTag('hiFJRhoProducer','mapEtaEdges'),
+                                 mapToRho = cms.InputTag('hiFJRhoProducer','mapToRho'),
+                                 mapToRhoM = cms.InputTag('hiFJRhoProducer','mapToRhoM'),
+                                 pfCandSource = cms.InputTag('particleFlowTmp'),
+                                 jetSource = cms.InputTag('kt4PFJets'),
+								 doCentrality = cms.untracked.bool(True),
+								 hiBinCut = cms.untracked.int32(100),    
+								 CentralityBinSrc = cms.InputTag("centralityBin","HFtowers"),
+								 keepGridInfo = cms.untracked.bool(False),
```
Comment (cvuosalo) [inline RecoHI/HiJetAlgos/python/hiFJGridEmptyAreaCalculator_cff.py]: Do not use untracked parameters for parameters that affect physics quantities and results. It looks like all these untracked parameters should be changed, including those in hiFJRhoProducer.py.

File: RecoHI/HiJetAlgos/plugins/HiFJGridEmptyAreaCalculator.cc [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+
+// ------------ method called once each job just after ending the event loop  ------------
+void 
+HiFJGridEmptyAreaCalculator::endJob() {
+}
+ 
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void HiFJGridEmptyAreaCalculator::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
```
Comment (cvuosalo) [inline RecoHI/HiJetAlgos/plugins/HiFJGridEmptyAreaCalculator.cc]: Please follow these instructions.

File: RecoHI/HiJetAlgos/plugins/HiFJRhoProducer.cc [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+// ------------ method called once each job just after ending the event loop  ------------
+void 
+HiFJRhoProducer::endJob() {
+}
+ 
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void HiFJRhoProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
```
Comment (cvuosalo) [inline RecoHI/HiJetAlgos/plugins/HiFJRhoProducer.cc]: Please complete fillDescriptions as instructed.

File: RecoHI/HiJetAlgos/plugins/HiFJGridEmptyAreaCalculator.h [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+#include <vector>
+
+// user include files
+#include "FWCore/Framework/interface/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+
+#include "TMath.h"
```
Comment (cvuosalo) [inline RecoHI/HiJetAlgos/plugins/HiFJGridEmptyAreaCalculator.h]: Please don't use TMath. Use std instead. Change TMath.Pi() to `M_PI`.

File: RecoHI/HiJetAlgos/plugins/HiFJGridEmptyAreaCalculator.h [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  ///input parameters
+  double gridWidth_, band_;
+  int hiBinCut_;
+  bool doCentrality_;
+  bool keepGridInfo_;
+  
+  std::vector<double> _rho_vs_eta;
+  std::vector<double> _mean_rho_vs_eta;
+  std::vector<double> _eta_max_grid;
+  std::vector<double> _eta_min_grid;
```
Comment (cvuosalo) [inline RecoHI/HiJetAlgos/plugins/HiFJGridEmptyAreaCalculator.h]: Please obey the CMS naming conventions: no underbars in method names, no leading underbars, member names use trailing underbars. So:
```
calculate_grid_rho -> calculateGridRho
twopi -> twoPi
_ymin -> ymin_
etc.
```
Please fix the names in all files in this PR as necessary.

Rules are here: https://cms-cpt-software.web.cern.ch/cms-cpt-software/General/CodingAndStyleRules.pdf

**Comment by cvuosalo** [comment]
@dgulhan -- There are several static analyzer errors in HiFJRhoproducer.h/.cc that need to be fixed:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14985/13756/llvm-analysis/

File: RecoHI/HiJetAlgos/plugins/HiFJGridEmptyAreaCalculator.cc [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+using namespace edm;
+
+
+HiFJGridEmptyAreaCalculator::HiFJGridEmptyAreaCalculator(const edm::ParameterSet& iConfig):
+  gridWidth_(iConfig.getUntrackedParameter<double>("gridWidth",0.005)),
+  band_(iConfig.getUntrackedParameter<double>("bandWidth",0.2)),
+  hiBinCut_(iConfig.getUntrackedParameter<int>("hiBinCut",60)),
+  doCentrality_(iConfig.getUntrackedParameter<bool>("doCentrality",true)),
+  keepGridInfo_(iConfig.getUntrackedParameter<bool>("keepGridInfo",true))
+{
```
Comment (cvuosalo) [inline RecoHI/HiJetAlgos/plugins/HiFJGridEmptyAreaCalculator.cc]: Please initialize all numerical data members in the constructor.

**Comment by dgulhan** [comment]
Thanks to @mverwe PR comments are addressed following the instructions we received. We also updated the customise_PPwithHI.py following the comments in the meeting to pick up changes of PR 15048. Is there anything else to be done?

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@dgulhan as I mentioned last Wednesday, the names of customization functions have changed in #15056 , which is merged now.
You will need to update this PR to resolve merge conflicts in  RecoHI/Configuration/python/customise_PPwithHI.py

**Comment by cvuosalo** [comment]
@dgulhan: Ping. When will you update this PR?

**Comment by dgulhan** [comment]
@slava77  @cvuosalo  I had changed the names of those functions already with this update to the branch:
https://github.com/cms-sw/cmssw/pull/14985/commits/9d1f0c1cfaed32b92a1b511c4f362703ed78f448
I thought the tests failed due to file open errors and I missed the comment by @slava77 from 7 days ago.

**Comment by slava77** [comment]
@dgulhan 
you can see "This branch has conflicts that must be resolved" message in this PR web page.
An update is needed for this to go away.

I'd suggest to rebase, but you have merge commits. So, it's unclear if it will work out.
http://cms-sw.github.io/tutorial-forward-port.html or another way to get this PR in sync with the latest 81X will need to be done

**Comment by cvuosalo** [comment]
@dgulhan: When might you be able to resolve the conflicts in this PR?

**Comment by dgulhan** [comment]
@cvuosalo sorry for delays, I made a new PR:
https://github.com/cms-sw/cmssw/pull/15233
I am closing this one.

**Pull request closed without merging**